File: Day213_2025/Q213_GFG.cpp
class Solution {
    bool isVowel(char ch) {
        return ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u';
    }
  public:
    int countBalanced(vector<string>& arr) {
        // Q. Balancing Consonants and Vowels Ratio
        // code here
        unordered_map<int, int> balanceFreq;
        balanceFreq[0] = 1;

        int count = 0;
        int prefixBalance = 0;

        for (string& s : arr) {
            int bal = 0;
            for (char ch : s) {
                if (isVowel(ch)) bal++;
                else bal--;
            }

            prefixBalance += bal;
            count += balanceFreq[prefixBalance];
            balanceFreq[prefixBalance]++;
        }

        return count;
    }
};

File: Day213_2025/Q213_GFG.py
from collections import defaultdict

class Solution:
    def countBalanced(self, arr):
        def is_vowel(ch):
            return ch in 'aeiou'
        
        balance_freq = defaultdict(int)
        balance_freq[0] = 1

        count = 0
        prefix_balance = 0

        for s in arr:
            bal = 0
            for ch in s:
                if is_vowel(ch):
                    bal += 1
                else:
                    bal -= 1

            prefix_balance += bal
            count += balance_freq[prefix_balance]
            balance_freq[prefix_balance] += 1

        return count
